sympy/polys/agca/extensions.py

```diff
@@ -109,12 +153,51 @@ def __str__(f):
 
 
 class MonogenicFiniteExtension(object):
-    """
+    r"""
     Finite extension generated by an integral element.
 
     The generator is defined by a monic univariate
     polynomial derived from the argument ``mod``.
 
+    A shorter alias is ``FiniteExtension``.
+
+    Examples
+    ========
+
+    Quadratic integer ring $\mathbb{Z}[\sqrt2]$:
+
+    >>> from sympy import Symbol, Poly
+    >>> from sympy.polys.agca.extensions import FiniteExtension
+    >>> x = Symbol('x')
+    >>> R = FiniteExtension(Poly(x**2 - 2)); R
+    ZZ[x]/(x**2 - 2)
+    >>> R.rank
+    2
+    >>> R(1 + x)*(3 - 2*x)
+    x - 1
+
+    Finite field $GF(5^3)$ defined by the primitive
+    polynomial $x^3 + x^2 + 2$ (over $\mathbb{Z}_5$).
+
+    >>> F = FiniteExtension(Poly(x**3 + x**2 + 2, modulus=5)); F
+    GF(5)[x]/(x**3 + x**2 + 2)
+    >>> F.basis
+    (1, x, x**2)
+    >>> F(x + 3)/(x**2 + 2)
+    -2*x**2 + x + 2
+
+    Function field of an elliptic curve:
+
+    >>> t = Symbol('t')
+    >>> FiniteExtension(Poly(t**2 - x**3 - x + 1, t, field=True))
+    ZZ(x)[t]/(t**2 - x**3 - x + 1)
+
+    Notes
+    =====
+
+    It's not possible to use a ``FiniteExtension`` as ``domain`` for the
+    :class:`~.Poly` class.
```

Comment:
Why can't this be used as a domain?

Reply:
A more a fitting comment might be that `FiniteExtension` is not a subclass of `Domain`. However, it should be possible to define a `FiniteField` class that would derive from `Domain` and `FiniteExtension`. It is also possible to define other wrapper subclasses of `Domain` for (some types of)  `FiniteExtension`. (Consider `PolynomialRing` vs. `PolyRing`.)

Reply:
> A more a fitting comment might be that `FiniteExtension` is not a subclass of `Domain`. However, it should be possible to define a `FiniteField` class that would derive from `Domain` and `FiniteExtension`. It is also possible to define other wrapper subclasses of `Domain` for (some types of) `FiniteExtension`. (Consider `PolynomialRing` vs. `PolyRing`.)

I'll see what this entails and how this works out and at the very least improve the comment.

Finite fields were the motivation for looking into this, but in the meantime I've also had a closer look at `galoistools.py` and it definitely has everything that would be needed for implementing proper non-prime finite fields (including algorithms for determining irreducibility over Z_p and actually finding irreducible polynomials). So maybe that's the better way forward when it comes to implementing GF(p^n) as polys domain. At least it seems to have been the original plan.

Reply:
I'm trying to work out a proof-of-concept for GF(p**d) using FiniteExtension, but I must confess that I'm struggling a bit with all the (messy/undocumented/labyrinthine) internals of `polys`. In particular:
- what is the `lev` argument of `DMP`?
https://github.com/sympy/sympy/blob/a69a68cf278728663c9ced8a9b1a1cb4dfc6355a/sympy/polys/polyclasses.py#L145-L162
- I noticed the following:
```python
In [60]: F = FiniteExtension(poly(x**3+x**2+2, modulus=5))

In [61]: type(F.ring)
Out[61]: sympy.polys.domains.old_polynomialring.GlobalPolynomialRing
```
Isn't the `old_polynomialring` somewhat deprecated? Does it matter for this application?

Reply:
I think the lev argument is basically the number of generators:
```julia
In [5]: Poly(x, x, domain=ZZ).rep.lev                                                                                                          
Out[5]: 0

In [6]: Poly(x, x, y, domain=ZZ).rep.lev                                                                                                       
Out[6]: 1

In [7]: Poly(x, x, y, z, domain=ZZ).rep.lev                                                                                                    
Out[7]: 2
```
I think it's called `lev` as short for "level" representing the fact that in the dense representation a poly in `n` generators is implemented as a poly of polys in `n-1` generators. The internal representation is a list of domain elements for an univariate poly and then a list of lists for bivariate etc:
```julia
In [9]: Poly(x, x, domain=ZZ).rep.rep                                                                                                          
Out[9]: [mpz(1), mpz(0)]

In [10]: Poly(x, x, y, domain=ZZ).rep.rep                                                                                                      
Out[10]: [[mpz(1)], []]

In [11]: Poly(x, x, y, z, domain=ZZ).rep.rep                                                                                                   
Out[11]: [[[mpz(1)]], [[]]]
```
The `lev` argument basically tells the `dmp_*` functions how many levels to recurse.

Reply:
> I think the lev argument is basically the number of generators [...]

Thanks!

I'm also hitting circular import issues (concerning `Poly`) when I try adding a `PrimePowerFiniteField` class in `domains/finitefield.py` that subclasses `FiniteField` and `FiniteExtension`. That's probaby not unexpected given that logically the chain is domains -> polynomials -> extensions. Maybe a more direct implementation is less prone to such issues.

Concerning this PR I'll modify the note in the docstring and that's it.

Reply:
> Isn't the old_polynomialring somewhat deprecated?

It depends on the number of generators. For a small number of generators, say, 1, 2, (3,) and dense polynomials, I think that it is quite efficient. For a large number of generators, the nested list representation becomes very inefficient.

Reply:
`lev` is the number of generators minus 1.

Reply:
> 'm also hitting circular import issues (concerning Poly) when I try adding a PrimePowerFiniteField class in domains/finitefield.py that subclasses FiniteField and FiniteExtension.

I have suggested another approach: Define a new class, say, `FiniteRing`, that would do what `FiniteField` does now (or simply rename `FiniteField`). Invoking `FiniteField` with a non-prime argument would pass the call to `FiniteRing`, presumably with a deprecation statement. Then `FiniteField` could subclass from `FiniteExtension`.

Reply:
> presumably with a deprecation statement

I don't think that's necessary as long as the new `FiniteField` also accepts the current parameters for creating a finite field of prime order. There are only some 3 or 4 very minor (and non-sensical) tests in the whole testsuite that actually instantiate a ring of integers modulo a composite. There's no real use anywhere and the current broken behavior is (fortunately) not publicly documented.

I think it's best to take the discussion concerning the implementation to #9544.